src/components/Dropdown/Dropdown.tsx

```diff
@@ -0,0 +1,137 @@
+import PopperJS, { Data, ModifierFn } from 'popper.js';
+import React, { FunctionComponent, ReactNode, useState } from 'react';
+import { Manager, Popper, Reference } from 'react-popper';
+import { Icon } from '../..';
+import { useClickOutside } from '../../hooks/useClickOutside';
+import { useKeyPress } from '../../hooks/useKeyPress';
+
+export interface DropdownProps {
+	label: string;
+	placement?:
+		| 'auto'
+		| 'auto-start'
+		| 'auto-end'
+		| 'top'
+		| 'top-start'
+		| 'top-end'
+		| 'right'
+		| 'right-start'
+		| 'right-end'
+		| 'bottom'
+		| 'bottom-start'
+		| 'bottom-end'
+		| 'left'
+		| 'left-start'
+		| 'left-end';
+	autoSize?: boolean; // defaults to false, meaning: the flyout with will be the same as the reference element
+	children: ReactNode;
+}
+
+/**
+ * This component provides a button that can show a flyout with some children inside of it.
+ * The PopperJS library is used to provide the positioning logic for the flyout element.
+ *
+ * The nomenclature within this library is as follows:
+ * - The button with down arrow is called the "reference"
+ * - The flyout element that contains the children is called the "popper"
+ */
+export const Dropdown: FunctionComponent<DropdownProps> = ({
+	label,
+	placement = 'top-start',
+	autoSize = false,
+	children,
+}: DropdownProps) => {
+	const [isOpen, setOpen] = useState(false);
+
+	let dropdownButton: HTMLButtonElement | null = null;
+	let dropdownFlyout: HTMLDivElement | null = null;
+
+	/**
+	 * Check if passed element is part of the component
+	 * @param elem
+	 */
+	const isPartOfElement = (elem: Element | null): boolean => {
+		return Boolean(
+			elem &&
+				(!dropdownButton || !dropdownButton.contains(elem)) &&
+				(!dropdownFlyout || !dropdownFlyout.contains(elem))
+		);
+	};
+
+	/**
+	 * Toggles the flyout elements visibility if no boolean is passed
+	 * If you pass "true", the flyout will be set to visible, even if it was visible before the call
+	 * If you pass "false" the flyout will be hidden, even if it was hidden before the call
+	 * @param shouldOpen
+	 */
+	const toggleOpen = (shouldOpen?: boolean) => {
+		setOpen(!isUndefined(shouldOpen) ? shouldOpen : !isOpen);
+	};
+
+	/**
+	 * We let popper calculate all the required styles, then we modify them a little based on the autoSize settings
+	 * @param data
+	 * @param options
+	 */
+	const computeStyle = (data: Data, options: Object) => {
+		const computeStylesFn: ModifierFn = get(PopperJS, 'Defaults.modifiers.computeStyle.fn');
+		if (!computeStylesFn) {
+			// TODO show error
+			return data;
+		}
+		const newData = computeStylesFn(data, options);
+		if (!autoSize) {
+			// Make the width off the popper the same size as the reference element
+			newData.styles.width = `${newData.offsets.reference.width}px`;
+		}
+		newData.styles.overflow = 'hidden';
+		return newData;
+	};
+
+	const modifiers = {
+		computeStyle: {
+			fn: computeStyle,
+		},
+	};
+
+	useKeyPress('Escape', () => toggleOpen(false));
```

Comment:
Verplaats deze anders ook naar bovenaan, waar de `useState`-hook ook staat?

Reply:
isPartOfElement is defined as const function
dus die moet ervoor gedefinieerd worden
![image](https://user-images.githubusercontent.com/1710840/59410341-ea8b3000-8db8-11e9-847d-4560fc0e08e3.png)

ik kan wel de useState naar beneden verplaatsen?